File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/ride/RateDriverScreen.kt
package lk.mageride.passenger.ride

import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.outlined.StarBorder
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.LabelledTextField
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme

/**
 * SCR-PA-019 — *"How was your ride with K. Fernando?"*.
 *
 * Stars, the four compliment chips, an optional comment, Submit (US-18.1).
 *
 * **Submit saves rather than sends, and the copy says so**, because there is no contract to send
 * it to: `ride.yaml` declares no rating operation and trip-state-svc's is scoped to a *session*,
 * which a Mode C ride is not. See [RateDriverViewModel] for the whole argument and the C080
 * handoff for the route this needs. Telling a passenger their rating was submitted would be
 * telling them something that did not happen.
 */
@Composable
internal fun RateDriverScreen(onDone: () -> Unit, model: RateDriverViewModel) {
    val state by model.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.queued) { if (state.queued) onDone() }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(MageRideTheme.spacing.md),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.md),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            text = state.driverName
                ?.let { stringResource(R.string.rate_title, it) }
                ?: stringResource(R.string.rate_title_unknown),
            style = MaterialTheme.typography.titleLarge,
            color = MaterialTheme.colorScheme.onSurface,
            textAlign = TextAlign.Center,
        )

        Row(horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
            for (star in 1..RateDriverState.MAX_STARS) {
                Icon(
                    imageVector = if (star <= state.stars) Icons.Filled.Star else Icons.Outlined.StarBorder,
                    contentDescription = stringResource(R.string.rate_stars, star),
                    modifier = Modifier
                        .size(ControlTokens.Star)
                        .clickable { model.setStars(star) },
                    tint = if (star <= state.stars) {
                        MaterialTheme.colorScheme.primary
                    } else {
                        MaterialTheme.colorScheme.outline
                    },
                )
            }
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            RatingTag.entries.forEach { tag ->
                FilterChip(
                    selected = tag in state.tags,
                    onClick = { model.toggleTag(tag) },
                    label = { Text(stringResource(tag.label)) },
                )
            }
        }

        LabelledTextField(
            label = stringResource(R.string.rate_comment),
            value = state.comment,
            onValueChange = model::onCommentChanged,
            placeholder = stringResource(R.string.rate_comment_hint),
            keyboardType = KeyboardType.Text,
            imeAction = ImeAction.Done,
        )

        state.error?.let { InlineError(message = stringResource(it)) }

        MageRideCta(
            label = stringResource(R.string.rate_submit),
            onClick = model::submit,
            enabled = state.canSubmit,
            loading = state.submitting,
        )
    }
}

File: apps/passenger-ios/PassengerApp/Ride/RateDriverScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-019 — *"Rate K. Fernando"*.
///
/// The cell: a scrimmed map with a sheet over it, a large avatar, the driver's name, `★★★★☆`, the
/// compliment chips, a comment field and **Submit** (US-18.1). *"star tap scale + haptic"* is its own
/// state line and both live in ``StarRating``.
///
/// **Submit saves rather than sends, and the copy says so**: `ride.yaml` declares no rating operation
/// at all, and trip-state-svc's is scoped to a *session*, which a Mode C ride is not — calling it
/// with a ride id would cross R-01. See ``RideRatings``. Telling a passenger their rating was
/// submitted would be telling them something that did not happen.
@MainActor
struct RateDriverScreen: View {

    @StateObject private var model: RateDriverModel

    /// Saved. Back to wherever the passenger came from — the receipt, or the map.
    let onDone: () -> Void

    init(
        rideId: String,
        rides: RideRepository,
        ratings: RideRatings,
        onDone: @escaping () -> Void
    ) {
        _model = StateObject(wrappedValue: RateDriverModel(rideId: rideId, rides: rides, ratings: ratings))
        self.onDone = onDone
    }

    var body: some View {
        ScrimmedSheet {
            DriverIdentityRow(name: model.state.driverName, size: MageRideControl.avatarLarge)
                .frame(maxWidth: .infinity)

            StarRating(value: model.state.stars) { model.setStars($0) }
                .frame(maxWidth: .infinity)

            tags

            LabelledTextField(
                labelKey: "rate_comment",
                value: commentBinding,
                placeholder: "rate_comment_hint".localised,
                autocapitalisation: .sentences
            )

            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
            }

            Button { model.submit() } label: { Text(key: "rate_submit") }
                .buttonStyle(.mageCta(loading: model.state.isSubmitting))
                .disabled(!model.state.canSubmit)
        }
        .toolbar(.hidden, for: .navigationBar)
        .task { model.start() }
        .onChange(of: model.state.isQueued) { queued in
            if queued { onDone() }
        }
    }

    /// The cell's wrapping chip row.
    ///
    /// A `FlowRow`-style `Layout` would be the exact CSS, and this app has none; two rows of two is
    /// the same shape at every Dynamic Type size and needs no custom layout — which matters here,
    /// because the four labels are longest in Sinhala and Tamil.
    private var tags: some View {
        VStack(spacing: MageRideSpacing.xs) {
            ForEach(RateDriverScreen.tagRows, id: \.first) { row in
                HStack(spacing: MageRideSpacing.xs) {
                    ForEach(row, id: \.self) { tag in
                        PlaceChip(
                            label: tag.labelKey.localised,
                            symbolName: model.state.tags.contains(tag) ? "checkmark" : "hand.thumbsup"
                        ) {
                            model.toggle(tag)
                        }
                        .opacity(model.state.tags.contains(tag) ? 1 : 0.6)
                    }
                }
            }
        }
        .frame(maxWidth: .infinity)
    }

    private var commentBinding: Binding<String> {
        Binding(
            get: { model.state.comment },
            set: { model.onCommentChanged($0) }
        )
    }

    /// The four tags, two to a row.
    ///
    /// Keyed by `first` in the `ForEach` above rather than by index: Swift has no key paths into
    /// tuples and indexing a collection by position reorders it the moment the data changes — the
    /// finding `apps/driver-ios` records from C087 onwards.
    private static let tagRows: [[RatingTag]] = [
        [.clean, .onTime],
        [.polite, .safeDriving],
    ]
}
